File: usuario-api/src/main/java/com/example/usuario_api/service/ProductoService.java
// src/main/java/com/example/usuario_api/service/ProductoService.java
package com.example.usuario_api.service;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import com.example.usuario_api.model.Producto;
import com.example.usuario_api.repository.ProductoRepository;

@Service
public class ProductoService {
    private final ProductoRepository repo;

    public ProductoService(ProductoRepository repo) {
        this.repo = repo;
    }

    public Page<Producto> listar(
    Optional<Long> categoriaId,
    Optional<BigDecimal> minPrecio,
    Optional<BigDecimal> maxPrecio,
    Pageable pageable
) {
    BigDecimal min = minPrecio.orElse(BigDecimal.ZERO);
    BigDecimal max = maxPrecio.orElse(BigDecimal.valueOf(Double.MAX_VALUE));

    if (categoriaId.isPresent()) {
        return repo.findByCategoriaIdAndPrecioBetween(
            categoriaId.get(), min, max, pageable);
    } else {
        return repo.findByPrecioBetween(min, max, pageable); // ✅ Nuevo método
    }
}

    public Optional<Producto> porId(Long id) {
        return repo.findById(id);
    }

    // Obtener los 4 productos con mayor descuento (ofertas del día)
    public List<Producto> obtenerOfertasDelDia() {
        List<Producto> productos = repo.findTop4ByDescuentoOrderByPorcentajeDesc();
        return productos.stream().limit(4).collect(Collectors.toList());
    }

    // Obtener 5 productos con descuento que no tengan más de una semana
    public List<Producto> obtenerOfertasDeLaSemana() {
        LocalDate fechaLimite = LocalDate.now().minusWeeks(1);
        List<Producto> productos = repo.findTop5ByDescuentoAndFechaCreacionAfterOrderByPorcentajeDesc(fechaLimite);
        return productos.stream().limit(5).collect(Collectors.toList());
    }

    // Obtener productos más vendidos (máximo 4)
    public List<Producto> obtenerProductosMasVendidos() {
        List<Object[]> resultados = repo.findProductosMasVendidos();
        return resultados.stream()
                .limit(4)
                .map(resultado -> (Producto) resultado[0])
                .collect(Collectors.toList());
    }

    // Obtener productos nuevos (menos de una semana)
    public List<Producto> obtenerProductosNuevos() {
        LocalDate fechaLimite = LocalDate.now().minusWeeks(1);
        return repo.findProductosNuevos(fechaLimite);
    }
}

File: user-frontend/src/services/productoService.js
const API_BASE_URL = 'http://192.168.18.2:8080/api/usuario';

export const productoService = {
  // Obtener productos con mayor descuento (máximo 4)
  async getOfertasDia() {
    try {
      const response = await fetch(`${API_BASE_URL}/productos/ofertas-dia`);
      if (!response.ok) {
        throw new Error('Error al obtener ofertas del día');
      }
      return await response.json();
    } catch (error) {
      console.error('Error en getOfertasDia:', error);
      return [];
    }
  },

  // Obtener productos con mayor descuento de la semana (máximo 5)
  async getOfertasSemana() {
    try {
      const response = await fetch(`${API_BASE_URL}/productos/ofertas-semana`);
      if (!response.ok) {
        throw new Error('Error al obtener ofertas de la semana');
      }
      return await response.json();
    } catch (error) {
      console.error('Error en getOfertasSemana:', error);
      return [];
    }
  },

  // Obtener productos más vendidos (máximo 4)
  async getMasVendidos() {
    try {
      const response = await fetch(`${API_BASE_URL}/productos/mas-vendidos`);
      if (!response.ok) {
        throw new Error('Error al obtener productos más vendidos');
      }
      return await response.json();
    } catch (error) {
      console.error('Error en getMasVendidos:', error);
      return [];
    }
  },

  // Obtener productos nuevos (menos de una semana)
  async getProductosNuevos() {
    try {
      const response = await fetch(`${API_BASE_URL}/productos/nuevos`);
      if (!response.ok) {
        throw new Error('Error al obtener productos nuevos');
      }
      return await response.json();
    } catch (error) {
      console.error('Error en getProductosNuevos:', error);
      return [];
    }
  }
}; 